### PR TITLE
Simultaneous model-based fMRI

### PR DESCRIPTION
**Pull request by mshvartsman**
Here is a writeup of simultaneous model-based fMRI using TFA for dimension reduction and emcee for MCMC estimation. It's not quite 100% working but I hope that this pull request can serve as a place to have discussion and collaboration in hopes of getting it ready for brainiak.

### CODE REVIEW

**Comment by buildbot-princeton** [comment]
Can one of the admins verify this patch?

**Comment by buildbot-princeton** [comment]
Can one of the admins verify this patch?

**Comment by mihaic** [comment]
Add to whitelist.

File: examples/smb-fMRI.ipynb [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "from nilearn.input_data import NiftiMasker\n",
+    "import numpy as np\n",
+    "import pandas as pd\n",
```
Comment (mihaic) [inline examples/smb-fMRI.ipynb]: You need to add all new dependencies to `requirements.txt`.

File: examples/smb-fMRI.ipynb [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    "    x0, t0, cong_a, incong_a, cong_z, incong_z = wfptParamTransform(x0, t0, cong_a, incong_a, cong_z, incong_z)\n",
+    "\n",
+    "    logger.debug(\"behav params: x0=%f, t0=%f, cong_a=%f, incong_a=%f, cong_z=%f, incong_z=%f\" % (x0, t0, cong_a, incong_a, cong_z, incong_z))\n",
+    "\n",
+    "    # 5 TFA centers so neuroparams will have a length-5 vector\n",
+    "    neuromean = np.array(neuroparams[0:5])\n",
+    "    # next, cholesky of 5x5 covariance matrix (to ensure positive semidefinite)\n",
+    "    neuroChol = np.zeros((5, 5))\n",
+    "    neuroChol[np.triu_indices(5, 0)] = np.array(neuroparams[5:20])\n",
+    "    neuroCov = neuroChol.T @ neuroChol\n",
```
Comment (mihaic) [inline examples/smb-fMRI.ipynb]: We still support Python 3.4, so we cannot use `@`.

File: examples/smb-fMRI.ipynb [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+    "\n",
+    "In the Turner et al. papers, the dimension reduction is done independently, so they use point estimates of the features rather than distirbutions. Conveniently this means we can directly reuse feature extraction pipelines like TFA for doing this off-the-shelf. That said, the right way to do this would be in a semi-supervised way where the dimension reduction is informed by the rest of the model. \n",
+    "\n",
+    "We apply this here to the NYU slow event-related flanker task, which allows us to use wiener first passage time distributions to model behavioral response times and choices. The dataset is available at https://openfmri.org/dataset/ds000102/, though ideally for this example we would find a dataset in the examples provided with some existing package (but is has to be a simple behavioral task like random dots, stroop/flanker, etc). \n",
```
Comment (mihaic) [inline examples/smb-fMRI.ipynb]: You should download the data from the Python code.

File: examples/smb-fMRI.ipynb [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    "    img = nib.load(nifti_file)\n",
+    "    all_images = np.float64(nifti_masker.fit_transform(img))\n",
+    "    data = all_images.T.copy()\n",
+    "\n",
+    "    #save data\n",
+    "    return data, R\n",
+    "\n",
+    "\n",
+    "# get brain data\n",
+    "neurodata1,R1 = extract_data('PATH_TO_DOWNLOADED_DATA.nii.gz')\n",
```
Comment (mihaic) [inline examples/smb-fMRI.ipynb]: Later you use `neuradata` and `R` (without the `1` suffix). Please correct.

**Comment by TuKo** [comment]
Hi Mike,

I ran the notebook with
ds102_R2.0.0/sub-01/func/sub-01_task-flanker_run-1_bold.nii.gz (and its tsv event file)
and I'm getting a singular matrix error while running mcmc. 

See the details below.

Javier

params: [ 1.37 -2.23 -0.22 -0.67  0.75  1.07  0.01  0.2   2.61 -0.34 -0.7   4.61
 -2.87  1.2   1.15 -0.22  1.42 -0.36 -1.86 -0.46  3.2   0.67 -2.53  0.74
  0.11  1.37 -0.28  0.68  1.73  0.69  0.72  0.54  0.24  0.68  1.11  0.69
  0.15  0.4  -2.44 -0.85  2.03  0.69 -0.86 -1.66 -0.19  1.62  3.03  0.57
  1.67 -0.69 -0.4   1.41 -0.12 -0.14 -0.02  1.12 -0.09 -0.04  1.26 -0.01
  0.65 -0.4   2.29  0.91 -0.88 -1.48  0.02  1.92 -1.27  0.34 -0.99 -0.13
 -2.6   0.04 -0.05  1.09  1.37 -0.51 -0.26  1.4   2.18  0.02  0.94 -0.69
 -1.51  0.59 -0.41  0.86  1.72 -0.95  0.25 -0.43  0.95  0.12  0.14  0.96
 -0.9   0.86 -0.64  0.57 -0.04 -2.25  1.22]
  args: (array([[ -7.69,   8.27,  10.86,  -7.28,  -9.22],
       [  7.89,  -8.75,  10.48,   8.57,  -7.71],
       [ -7.21,   0.15,  -0.01,   7.39,   0.06]]), array([[ 1.09,  0.99,  0.59,  0.5 ,  0.72,  0.54,  0.44,  0.47,  0.41,
         0.56,  0.49,  0.4 ,  0.47,  0.52,  0.56,  0.53,  0.44,  0.46,
         0.73,  0.48,  0.54,  0.54,  0.62,  0.49],
       [ 1.  ,  1.  ,  1.  ,  1.  ,  1.  ,  1.  ,  1.  ,  1.  ,  1.  ,
         1.  ,  1.  ,  1.  ,  1.  ,  1.  ,  1.  ,  1.  ,  1.  ,  1.  ,
         1.  ,  1.  ,  1.  ,  1.  ,  1.  ,  1.  ],
       [ 2.  ,  2.  ,  1.  ,  1.  ,  2.  ,  1.  ,  1.  ,  2.  ,  1.  ,
         2.  ,  1.  ,  1.  ,  1.  ,  2.  ,  2.  ,  2.  ,  1.  ,  1.  ,
         2.  ,  2.  ,  2.  ,  1.  ,  2.  ,  1.  ]]))

**Comment by mihaic** [comment]
Let's talk tomorrow at the hackathon about the possibility of making this code a module instead of an example.

File: examples/smb-fMRI.ipynb [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    "    # go from infinite support to where we actually want to be\n",
+    "    t0 = np.exp(t0)\n",
+    "    cong_a = np.exp(cong_a)\n",
+    "    incong_a = np.exp(incong_a)\n",
+    "    cong_z = np.exp(cong_z)\n",
+    "    incong_z = np.exp(incong_z)\n",
+    "    return x0, t0, cong_a, incong_a, cong_z, incong_z\n",
+    "\n",
+    "def wfptParamUntransform(x0, t0, cong_a, incong_a, cong_z, incong_z):\n",
+    "    # go from infinite support to where we actually want to be\n",
```
Comment (TuKo) [inline examples/smb-fMRI.ipynb]: Please, correct the comment here. Is the same as above but the function is doing the opposite.

**Comment by TuKo** [comment]
@mshvartsman why do you add it as an example and not as a slightly more general implementation of the method for any function (and still have the example with TFA?

**Comment by TuKo** [comment]
An additional comment: besides the exception above, there is no output in the notebook.
Can you consider showing what you are doing visually? Like plotting the convergence, the results from MCMC in some way, or anything else that you find interesting. Thanks!

**Comment by mshvartsman** [comment]
Happy to discuss making this a module instead of an example tomorrow. I thought of it as an example because there's nothing novel here algorithm-wise: it's just putting together some pieces that already exist (TFA or ICA for dimension reduction, emcee for MCMC sampling, etc) with a pretty simple generative model.

**Comment by mshvartsman** [comment]
My math is totally wrong, I misread the variable indices in the original paper, so I'm closing this until I do it right. I will take the comments given so far into account when I reopen.

**Pull request closed without merging**